File: src/lidar_lite.cpp
#include <include/lidar_lite.h>

using namespace std;

Lidar_Lite::Lidar_Lite (int bus){
  err = 0;
  adapter_num = bus;
  snprintf(filename, 19, "/dev/i2c-%d", adapter_num);
}

Lidar_Lite::~Lidar_Lite(void){
  printf("Ending Lidar-Lite Session");
  if (i2c_bus > 0){
   int e = close(i2c_bus);
  }
}

int Lidar_Lite::connect( void ) {
  printf("Connecting to %s", filename);
  i2c_bus = open(filename, O_RDWR);
  if (i2c_bus < 0){
    err = errno;
    printf("Connect Error: %d", err);
    return -1;
  }
  if (ioctl(i2c_bus, I2C_SLAVE, 0x62) < 0) {
    err = errno;
    printf("Bus Error: %d", err);
    return -1;
  }
  return 0;
}


int Lidar_Lite::writeAndWait(int writeRegister, int value){
  res = i2c_smbus_write_byte_data(i2c_bus, writeRegister, value);
  usleep(10000);
  if (res < 0){
    err = errno;
    printf("Write Error %d", err);
    return -1;
  } else {
    return 0;
  }
}

int Lidar_Lite::readAndWait(int readRegister){
  res = i2c_smbus_read_byte_data(i2c_bus, readRegister);
  usleep(10000);
  if (res < 0){
    err = errno;
    printf("Read Error: %d", err);
    return -1;
  } else {
    return 0;
  }
}

int Lidar_Lite::getDistance( void ){
  int buf[2];
  int e = 0;
  e = writeAndWait(0x00,0x04);
  if (e < 0){
    return e;
  }
  e = readAndWait(0x8f);
  if (e < 0){
    return e;
  } else {
    buf[0] = res;
  }
  e = readAndWait(0x10);
  if (e < 0){
    return e;
  } else {
    buf[1] = res;
  }
  return (buf[0] << 8) + buf[1];
}

int Lidar_Lite::getError(void){
  return err;
}

File: python/lidar_lite.py
import smbus
import time

class Lidar_Lite():
  def __init__(self):
    self.address = 0x62
    self.distWriteReg = 0x00
    self.distWriteVal = 0x04
    self.distReadReg1 = 0x8f
    self.distReadReg2 = 0x10
    self.velWriteReg = 0x04
    self.velWriteVal = 0x08
    self.velReadReg = 0x09

  def connect(self, bus):
    try:
      self.bus = smbus.SMBus(bus)
      time.sleep(0.5)
      return 0
    except:
      return -1

  def writeAndWait(self, register, value):
    self.bus.write_byte_data(self.address, register, value);
    time.sleep(0.02)

  def readAndWait(self, register):
    res = self.bus.read_byte_data(self.address, register)
    time.sleep(0.02)
    return res

  def readDistAndWait(self, register):
    res = self.bus.read_i2c_block_data(self.address, register, 2)
    time.sleep(0.02)
    return (res[0] << 8 | res[1])

  def getDistance(self):
    try:
        self.writeAndWait(self.distWriteReg, self.distWriteVal)
        dist = self.readDistAndWait(self.distReadReg1)
    except (IOError, OSError) as err:
        print(err)
        return -1
        
    return dist

  def getVelocity(self):
    self.writeAndWait(self.distWriteReg, self.distWriteVal)
    self.writeAndWait(self.velWriteReg, self.velWriteVal)
    vel = self.readAndWait(self.velReadReg)
    return self.signedInt(vel)

  def signedInt(self, value):
    if value > 127:
      return (256-value) * (-1)
    else:
      return value


